### DISCUSSION TITLE
Fail fast vs fail safe

### DISCUSSION THREAD

**Discussion by waliahimanshu** [MobileNativeFoundation/discussions#80] [General]
Hi Folks, 

Curious about the opinions of crashing the app as early as possible as opposed to consuming errors and displaying an error view?

Throwing exceptions and crashing the app has the advantage of finding errors early, during dev/QA testing or worst case in the initial rollout. However, if a crash sneaks to the wider user group, the app becomes unusable. Typical examples are API contract violations or error in domain logic.

On the other hand, the fail-safe approach shows a relevant error screen with some parts of the app functional. But, actual errors are left unnoticed for long as they are disguised as non-fatal. These errors consume time in debugging sometimes involving other teams.
Some are more inclined to avoid a crash at all, as app stores tend to rank apps based on the crash rate.

How do you decide the fine line here?

**Comment by keith** [comment]
Not to say our approach is particularly novel, but we have some custom assertion functions that crash in dev / QA, and log to our crash reporter without crashing in production, and those logs are triaged by our mobile on calls. We're generally pretty conservative and use those everywhere instead of something that could actually crash in production, which likely leads to a slightly broken experience in some cases, but that's the side we've chosen.

**Reply by waliahimanshu** [reply]
Interesting to know the usage of custom assertions. This is something we have not explored and looks like a nice solution.
We too have taken a fail-safe approach where every error is handled gracefully by an error screen, but due to lack of on-call and some other gaps, we faced some issues lately.

Thanks for sharing this.

**Comment by vfn** [comment]
Some 3rd party crash reporting tools (Bugsnag, Firebase Crashlytics, &c.) support the logging of non-fatal errors. That could be used to track the issue while preventing the app from crashing if the degraded experience is acceptable. In some cases you might just want to let the app crash, but if you can handle the error "gracefully" while still capturing info that will allow you to debug it, why not?

Ive been using these to track a variety of assertions. On the setup I have in place, while in dev mode and connected to the debugger, the custom assertions will just pause the code (via `SIGINT`) and allow me to debug the problem without crashing the app. For release versions those custom assertions will then log to the registered handler that forwards the stack trace and custom info associated with the non-fatal error.

When debugging, I prefer to raise a `SIGINT` instead of crashing as that gives me the choice of whether to debug the issue as it happens or to capture some info and investigate it later, in the case I'm in the middle of something else more important.

**Reply by waliahimanshu** [reply]
Thanks for sharing this. Interesting to know the custom assertions and use of `SIGINT`. Looking forward to exploring this approach!
